### PR TITLE
Feat/api edit employee general info

### PR DESCRIPTION
**Pull request by datphamcode295**
#### What's this PR do?

- [x] API edit employee general info
- [x] Add column line_manager_id in employee table

#### What are the relevant Git tickets?
https://www.notion.so/API-to-edit-employee-info-62f923ea04a843ad96794cd1d0d3cb50

#### Screenshots (if appropriate)

// Use [Licecap](http://www.cockos.com/licecap/) to share a screencast gif.

image
#### Request
```
curl 'http://localhost:8200/api/v1/employees/2655832e-f009-4b73-a535-64c3a22e558f/general-info'
-X PUT
-H 'accept: /'
-H 'accept-language: vi,en-US;q=0.9,en;q=0.8'
-H 'content-type: application/json'
-H 'origin: http://localhost:3000/'
-H 'referer: http://localhost:3000/'
-H 'sec-ch-ua: "Chromium";v="106", "Google Chrome";v="106", "Not;A=Brand";v="99"'
-H 'sec-ch-ua-mobile: ?0'
-H 'sec-ch-ua-platform: "macOS"'
-H 'sec-fetch-dest: empty'
-H 'sec-fetch-mode: cors'
-H 'sec-fetch-site: cross-site'
-H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36'
-H 'authorization:Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTkzMjExNDIsImlkIjoiMjY1NTgzMmUtZjAwOS00YjczLWE1MzUtNjRjM2EyMmU1NThmIiwiYXZhdGFyIjoiaHR0cHM6Ly9zMy1hcC1zb3V0aGVhc3QtMS5hbWF6b25hd3MuY29tL2ZvcnRyZXNzLWltYWdlcy81MTUzNTc0Njk1NjYzOTU1OTQ0LnBuZyIsImVtYWlsIjoidGhhbmhAZC5mb3VuZGF0aW9uIiwicGVybWlzc2lvbnMiOlsiZW1wbG95ZWVzLnJlYWQiXSwidXNlcl9pbmZvIjpudWxsfQ.GENGPEucSUrILN6tHDKxLMtj0M0REVMUPC7-XhDMpGM'
--data-raw '{
    "fullName":"Phạm Đức Thành",
    "email":"thanh123@d.foundation",
    "phone":"0123456788"
}'
--compressed
```

#### Response
```
{
    "data": {
        "id": "2655832e-f009-4b73-a535-64c3a22e558f",
        "createdAt": "2022-11-02T09:52:34.586566Z",
        "updatedAt": "2022-11-14T22:28:28.817302Z",
        "fullName": "Phạm Đức Thành",
        "displayName": "Thanh Pham",
        "teamEmail": "thanh@d.foundation",
        "personalEmail": "thanhpham123@gmail.com",
        "avatar": "https://s3-ap-southeast-1.amazonaws.com/fortress-images/5153574695663955944.png",
        "phoneNumber": "0123456788",
        "address": "Phan Huy Ich, Tan Binh District, Ho Chi Minh, Vietnam",
        "mbti": "INTJ-A",
        "gender": "Male",
        "horoscope": "Libra",
        "birthday": "1990-01-02T00:00:00Z",
        "status": "contractor",
        "joinedDate": null,
        "leftDate": null,
        "accountStatus": "active",
        "seniority": {
            "id": "39735742-829b-47f3-8f9d-daf0983914e5",
            "createdAt": "2022-11-07T09:50:25.714604Z",
            "updatedAt": "2022-11-07T09:50:25.714604Z",
            "name": "Principal",
            "code": "principal"
        },
        "chapter": {
            "id": "11ccffea-2cc9-4e98-9bef-3464dfe4dec8",
            "createdAt": "2022-11-07T09:50:25.714604Z",
            "updatedAt": "2022-11-07T09:50:25.714604Z",
            "name": "Web",
            "code": "web"
        },
        "lineManager": null,
        "positions": [
            {
                "id": "11ccffea-2cc9-4e98-9bef-3464dfe4dec8",
                "createdAt": "2022-11-07T09:50:25.714604Z",
                "updatedAt": "2022-11-07T09:50:25.714604Z",
                "name": "Frontend",
                "code": "frontend"
            },
            {
                "id": "d796884d-a8c4-4525-81e7-54a3b6099eac",
                "createdAt": "2022-11-07T09:50:25.714604Z",
                "updatedAt": "2022-11-07T09:50:25.714604Z",
                "name": "Backend",
                "code": "backend"
            }
        ],
        "projects": []
    }
}
```

### CODE REVIEW

File: pkg/model/employees.go [hunk 1]
```diff
@@ -24,47 +24,55 @@ type Employee struct {
...
-	Address                 string
-	MBTI                    string
-	Gender                  string
-	Horoscope               string
-	PassportPhotoFront      string
-	PassportPhotoBehind     string
-	IdentityCardPhotoFront  string
-	IdentityCardPhotoBehind string
-	DateOfBirth             *time.Time
+	FullName               string     `gorm:"column:full_name"`
```
Comment (namnhce) [inline pkg/model/employees.go:27]: I think we don't need to use all gorm tags for all attributes. Only use for the one which cannot be recognized by ORM

File: pkg/routes/v1.go [hunk 2]
```diff
@@ -18,8 +18,9 @@ func loadV1Routes(r *gin.Engine, h *handler.Handler, cfg *config.Config) {
 	v1.GET("/profile", mw.WithAuth, h.Employee.GetProfile)
 
 	// employees
-	v1.GET("/employees", mw.WithAuth, mw.WithPerm(cfg, "employees.read"), h.Employee.List)
+	v1.GET("/employees", mw.WithAuth, h.Employee.List)
 	v1.GET("/employees/:id", h.Employee.One)
+	v1.POST("/employees/:id/general-info", mw.WithAuth, h.Employee.UpdateGeneralInfo)
```
Comment (namnhce) [inline pkg/routes/v1.go:23]: PUT

File: pkg/routes/v1.go [hunk 3]
```diff
@@ -18,8 +18,9 @@ func loadV1Routes(r *gin.Engine, h *handler.Handler, cfg *config.Config) {
...
 	v1.POST("/employees/:id/employee-status", mw.WithAuth, h.Employee.UpdateEmployeeStatus)
```
Comment (namnhce) [inline pkg/routes/v1.go:24]: PUT

File: pkg/view/employee.go [hunk 4]
```diff
@@ -27,6 +27,12 @@ type EmployeeData struct {
 	WorkingStatus model.WorkingStatus `json:"status"`
 	JoinedDate    *time.Time          `json:"joinedDate"`
 	LeftDate      *time.Time          `json:"leftDate"`
+
+	AccountStatus model.AccountStatus `json:"accountStatus"`
+	Seniority     string              `json:"seniority"`
+	Chapter       string              `json:"chapter"`
+	LineManager   string              `json:"lineManager"`
```
Comment (namnhce) [inline pkg/view/employee.go:34]: This should be an object with ID Name and Avatar. and should be null if it's an empty object.

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/model/employees.go (hunk 1 above)
Comment (datphamcode295) [reply pkg/model/employees.go:27]: there are some field the gorm cannot recognized, so I used this solution for the whole structure to make things more consistent

File: pkg/handler/employee/testdata/list/have_working_and_no_pagination.json [hunk 5]
```diff
@@ -19,10 +19,29 @@
             "gender": "Male",
             "horoscope": "Libra",
             "birthday": "1990-01-02T00:00:00Z",
-            "status": "contractor",
```
Comment (namnhce) [inline pkg/handler/employee/testdata/list/have_working_and_no_pagination.json:22]: Why no status here

File: pkg/model/employees.go (hunk 1 above)
Comment (namnhce) [reply pkg/model/employees.go:27]: well we can consider to change it in another PR for all model

File: pkg/handler/employee/employee.go [hunk 6]
```diff
@@ -223,3 +223,58 @@ func (h *handler) GetProfile(c *gin.Context) {
...
+
+	// TODO: can we move this to middleware ?
+	l := h.logger.Fields(logger.Fields{
+		"handler": "employee",
+		"method":  "Update",
+		"request": body,
+	})
+
+	// 3. update informations and rerurn
+	employee, err := h.store.Employee.UpdateGeneralInfo(employee.EditGeneralInfo{
```
Comment (namnhce) [inline pkg/handler/employee/employee.go:265]: Check input employee exists.
Check lineManager exists.

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/employee/employee.go [hunk 7]
```diff
@@ -223,3 +223,59 @@ func (h *handler) GetProfile(c *gin.Context) {
...
+// @Summary Edit general info of the employee by id
+// @Description Edit general info of the employee by id
+// @Tags Employee
+// @Accept  json
+// @Produce  json
+// @Param id path string true "Employee ID"
+// @Param fullName body string true "fullName" maxlength(99)
+// @Param email body string true "email"
+// @Param phone body string true "phone" minlength(10)  maxlength(12)
+// @Param lineManager body model.UUID true "lineManager"
```
Comment (namnhce) [inline pkg/handler/employee/employee.go:237]:
```suggestion
// @Param lineManagerID body string true "lineManager"
```

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/store/employee/employee.go [hunk 8]
```diff
@@ -60,3 +60,18 @@ func (s *store) UpdateEmployeeStatus(employeeID string, accountStatus model.Acco
...
+	// 1.2 update infor
+	employee.FullName = body.Fullname
+	employee.TeamEmail = body.Email
+	employee.PhoneNumber = body.Phone
+	employee.LineManagerID = body.LineManagerID
+	employee.DiscordID = body.DiscordID
+	employee.GithubID = body.GithubID
+
+	// 1.3 save to DB
+	return employee, s.db.Table("employees").Where("id = ?", id).Updates(&employee).Preload("Chapter").Preload("Seniority").Preload("LineManager").Preload("EmployeePositions").Preload("EmployeePositions.Position").First(&employee).Error
```
Comment (namnhce) [inline pkg/store/employee/employee.go:79]: The limitation of character in a line should be less than 125.

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/store/employee/filter.go [hunk 9]
```diff
@@ -3,3 +3,13 @@ package employee
 type SearchFilter struct {
 	WorkingStatus string
 }
+
+type EditGeneralInfo struct {
+	Fullname      string `form:"fullName" json:"fullName" binding:"required,max=99"`
+	Email         string `form:"email" json:"email" binding:"required,email"`
```
Comment (namnhce) [inline pkg/store/employee/filter.go:9]:
```suggestion
	Fullname      string
	Email         string
```

**Review by namnhce — CHANGES_REQUESTED** [review]

**Review by namnhce — APPROVED** [review]

**Pull request merged**